src/Units/UnitModel.java

```diff
@@ -200,32 +202,49 @@ public boolean canRepairOn(Location locus)
   public GameEventQueue getTurnInitEvents(Unit self, MapMaster map)
   {
     GameEventQueue queue = new GameEventQueue();
-    if( isAirUnit() && 0 == self.fuel )
-    {
-      // Uh oh. We are about to go crashy crashy. Unless we are about to be resupplied...
-      boolean ok = false;
-      XYCoord xyc = new XYCoord(self.x, self.y);
-      Location loc = map.getLocation(xyc);
 
-      // First check if we are on an airport.
-      if(loc.getOwner() == self.CO && canRepairOn(loc))
-        ok = true;
+    XYCoord xyc = new XYCoord(self.x, self.y);
+    Location loc = map.getLocation(xyc);
+
+    // No actions for units in transports. Should also be checked in Unit.
+    if( null == loc ) return queue;
+
+    boolean resupplying = false;
+
+    // If the unit is not at max health, and is on a repair tile, heal it.
+    if( canRepairOn(loc) && !self.CO.isEnemy(loc.getOwner()) )
+    {
+      queue.add(new HealUnitEvent(self, self.CO.getRepairPower(), self.CO)); // Event handles cost logic
+      // Resupply is free; whether or not we can repair, go ahead and add the resupply event.
+      if( !self.isFullySupplied() )
+      {
+        resupplying = true;
+        queue.add(new ResupplyEvent(self, self));
+      }
+    }
 
-      // Next see if there is a friendly APC nearby.
+    // If there is an adjacent APC or similar, resupply.
+    if( !resupplying && !self.isFullySupplied() )
+    {
       List<XYCoord> adjacents = Utils.findLocationsInRange(map, xyc, 1);
       for( XYCoord adj : adjacents )
       {
         Unit res = map.getLocation(adj).getResident();
-        if( null != res && (res.CO == self.CO) && res.model.hasActionType(UnitActionFactory.RESUPPLY) )
+        if( (null != res) && !res.CO.isEnemy(self.CO) && res.model.hasActionType(UnitActionFactory.RESUPPLY) )
         {
-          ok = true;
+          queue.add(new ResupplyEvent(res, self));
+          resupplying = true;
           break;
         }
       }
+    }
 
-      if(!ok)
-        queue.add(new UnitDieEvent(self));
+    if( !resupplying && (0 == self.fuel) & (isAirUnit() || isSeaUnit()) )
```

Comment:
I still think it should be based on whether the unit has fuel drain, rather than unit type.

Reply:
But then land units would die. I'm pretty sure _that_ would not be cart-compliant.

Reply:
Cart *has* no land units with fuel drain.
I guess I could ask Xen/hack a ROM to figure out which case is "correct", but the carts frankly *don't address* that question at all.

Reply:
Hm. But why use `idleFuelBurn` instead of domain, since that seems like a more directly-related concept (falling out of the sky/sinking vs... randomly combusting once you are out of combustible materials).

Reply:
It's more a question of "needing the fuel on a continuous basis to function properly".
I don't pretend to know why naval units drain fuel in the way they do, though it makes perfect sense for air (though perhaps the relative quantities might want adjustment)

I'd honestly argue that infantry should have a fuel drain, since the only obvious "fuel" is food and you can't exactly fight on an empty stomach.

Reply:
"Needing fuel on a continuous basis to function" is a fair cry from "spontaneously explodes when out of fuel".

> naval units

It could be argues that it takes effort to maintain position with waves or whatever. I dunno.

> infantry

I'm not sure I want to add fuel drain to infantry, even if they don't explode when they run out. It would just be so annoying to try and resupply them all (not that most of them would live long enough for it to matter, but still).

Reply:
> "Needing fuel on a continuous basis to function" is a fair cry from "spontaneously explodes when out of fuel".

In the way that I meant it, it does.
You can turn your car engine off, and it's fine.
You can turn a jet's engine off, but you're probably in the air... and if so, you'll have "problems".
For sea, maybe it's something like a non-renewable factor in their power plants? If nuclear, some kind of stabilizing agent could conceivably be necessary to keep the thing from exploding, lol.

I still hold that passive fuel drain is the determining factor.

> infantry

It would be annoying, sure, but it would make perfect simulationist sense (which obviously isn't the same as perfect sense for the game).

Reply:
> I still hold that passive fuel drain is the determining factor.

And I hold that running out of fuel in a treacherous domain is the determining factor (plus it keeps us from having to make up strange justifications about stabilizing nuclear agents). I'll change it to `!isLandUnit()` if you want, but it doesn't make sense to base it on fuel drain.

> Simulationist Infantry

I still don't think I want that.

Reply:
> And I hold that running out of fuel in a treacherous domain is the determining factor

And what, pray tell, determines whether a domain is treacherous?
I would argue the bit of info that encodes that *is* the fuel drain - the unit has to expend effort to not-die on a continuous basis.
I would find it very curious if an air unit with no drain (say, a blimp or Eagle copters) exploded on using the last unit of fuel... but could chill indefinitely without the last bit of it.

Reply:
> I would find it very curious if an air unit with no drain (say, a blimp or Eagle copters) exploded on using the last unit of fuel... but could chill indefinitely with the last bit of it.

That's a fair point, though I'm not sure even blimps would count as having zero drain. Conversely, I'd find it very curious if a land unit with fuel drain (say if we wanted to limit BattleMech effectiveness or introduce a supply-challenged CO) exploded while just sitting there.

Perhaps the right answer is to use both: Units that don't sit on terra firma and do have fuel drain are susceptible.

Reply:
Bah. I did a quick test on Dual Strike, and Eagle's copters do indeed go boom IFF they're at exactly 0 fuel.
Go ahead and merge it, I guess.